File: azb.R
#!/usr/bin/env Rscript
# -*- coding: utf-8 -*-

#Created on Sun 05 Sep 2021 01:31:23 PM EDT
#author: Ryan Hildebrandt, github.com/ryancahildebrandt

# Doc Setup----
{


  library(DBI)
  library(duckdb)
  library(rvest)
  library(tidyverse)
  library(magrittr)
}

# Functions ----
get_text_html <- function(html_path, enc){
  out <- tryCatch(
    expr = {
      read_html(html_path, encoding = enc) %>%
        html_elements(., "body") %>%
        html_nodes(., "div.main_text") %>%
        html_text(.) %>%
        str_remove_all(., "\\n|\\r|\\s|[（《][ぁ-んァ-ンヽゞゝ／″＼]*[》）]")
    },
    error = function(e){
      message("Error:")
      print(e)
    }
  )
  out
}

get_text_zip <- function(zip, txt){
  unz(zip, txt) %>%
    read_lines(., locale = locale(encoding = "SHIFT-JIS")) %>%
    paste0(., collapse = "") %>%
    str_remove_all(., "\\n|\\r|\\s|[（《][ぁ-んァ-ンヽゞゝ／″＼｜]*[》）]|［＃.*?］|『.*?』")
}

main_text_clean <- function(txt){
  txt %>%
    str_remove_all(., "(---.*---)") %>%
    str_remove_all(., "底本[：・「:].*") %>%
    str_remove_all(., "入力(者)?[：・「:].*")
}

genres <- read_csv("./outputs/分類番号.csv")
genre_list <- deframe(genres)

# Readin ----
temp <- tempfile()
download.file("https://www.aozora.gr.jp/index_pages/list_person_all_extended_utf8.zip", temp)
azb_meta <- read_csv(unz(temp, "list_person_all_extended_utf8.csv"))
unlink(temp)

meta_df_raw <- azb_meta %>%
  rowwise(.) %>%
  mutate(.,
         html_path = str_replace(`XHTML/HTMLファイルURL`, "https://www.aozora.gr.jp", "./aozorabunko"),
         zip_path = str_replace(`テキストファイルURL`, "https://www.aozora.gr.jp", "./aozorabunko")) %>%
  filter(.,
         grepl("\\./aozorabunko", html_path),
         grepl("\\./aozorabunko", zip_path),
         file.exists(html_path),
         file.exists(zip_path)
  ) %>%
  mutate(.,
         main_text = list(get_text_html(html_path, `XHTML/HTMLファイル符号化方式`)),
         text_file = NA,
         text_length = length(main_text)
         )

html_success_df <- meta_df_raw %>%
  filter(., text_length == 1) %>%
  mutate(., main_text = unlist(main_text))

byte_error_df <- meta_df_raw %>%
  filter(., text_length > 1) %>%
  mutate(., text_file = list(unzip(zip_path, list = TRUE)$Name)) %>%
  mutate(., text_file = list(text_file[!is.na(str_extract(text_file, ".*\\.txt"))])) %>%
  mutate(., main_text = get_text_zip(zip_path, text_file))

zero_length_df <- meta_df_raw %>%
  filter(., text_length == 0) %>%
  mutate(., text_file = list(unzip(zip_path, list = TRUE)$Name)) %>%
  mutate(., main_text = get_text_zip(zip_path, text_file))

# Main dataframe compilation----
meta_df <- bind_rows(
  html_success_df,
  byte_error_df,
  zero_length_df)  %>%
  mutate(.,
         main_text = main_text_clean(main_text),
         分類番号 = gsub("[NDCK ]", "", 分類番号) %>% substr(., 1, 3),
         著者 = glue::glue(replace_na(名, ""), replace_na(姓, ""))) %>%
  mutate(.,
         n_char = nchar(main_text),
         分類 = purrr::map_chr(分類番号, ~genre_list[.x]),
         text_file = toString(text_file)
         ) %>%
  rowid_to_column(., "db_id")

en_cols <- c("db_id", "work_id", "work_name", "work_name_reading", "reading_sort", "subtitle", "subtitle_reading", "original_title", "first_appearance", "category_number", "character_type", "copyright_flag", "publication_date","last_updated", "card_url", "author_id","last_name", "first_name", "last_name_reading","first_name_reading", "last_name_reading_sort", "first_name_reading_sort","last_name_romaji", "first_name_romaji", "role_flag", "date_of_birth", "date_of_death", "personal_copyright_flag", "original_name_1", "original_publisher_1", "original_first_edition_publication_year_1", "input_version_1", "proofreading_version_1", "source_text_name_1", "source_text_publisher_1", "first_edition_publication_year_1", "original_name_2", "original_publisher_2", "original_first_edition_publication_year_2", "input_version_2", "proofreading_version_2", "source_text_name_2", "source_text_publisher_2", "first_edition_publication_year_2", "entered_by", "proofread_by", "text_file_url", "text_file_last_modified", "text_file_encoding", "text_file_character_set", "text_file_modification_count", "xhtml_html_file_url", "last_updated_xhtml_html_file", "xhtml_html_file_encoding", "xhtml_html_file_character_set", "xhtml_html_modification_count", "html_path", "zip_path", "main_text", "text_file", "text_length", "author", "n_char", "genre")
meta_df_en <- meta_df %>% set_names(., en_cols)

omitted_df <- azb_meta %>%
  filter(., !作品ID %in% meta_df_en$work_id) %>%
  rowwise(.) %>%
  mutate(.,
         html_path = str_replace(`XHTML/HTMLファイルURL`, "https://www.aozora.gr.jp", "./aozorabunko"),
         zip_path = str_replace(`テキストファイルURL`, "https://www.aozora.gr.jp", "./aozorabunko")
  )

# Export----
db_con <- dbConnect(
  duckdb(),
  dbdir = "./outputs/kaggle/aozora_corpus.db",
  )

authors_df <- meta_df_en %>%
  select(., author_id, author, last_name, first_name, last_name_reading, first_name_reading, last_name_reading_sort, first_name_reading_sort, last_name_romaji, first_name_romaji, date_of_birth, date_of_death, personal_copyright_flag) %>%
  unique(.)

works_df <- meta_df_en %>%
  select(., work_id, work_name, work_name_reading, reading_sort, subtitle, subtitle_reading, original_title, first_appearance, category_number, genre, character_type, copyright_flag, publication_date, last_updated, author_id, role_flag, original_name_1, original_publisher_1, original_first_edition_publication_year_1, input_version_1, proofreading_version_1, source_text_name_1, source_text_publisher_1, first_edition_publication_year_1, original_name_2, original_publisher_2, original_first_edition_publication_year_2, input_version_2, proofreading_version_2, source_text_name_2, source_text_publisher_2, first_edition_publication_year_2, entered_by, proofread_by) %>%
  unique(.)

texts_df <- meta_df_en %>%
  select(., work_id, text_file_url, text_file_last_modified, text_file_encoding, text_file_character_set, text_file_modification_count, xhtml_html_file_url, last_updated_xhtml_html_file, xhtml_html_file_encoding, xhtml_html_file_character_set, xhtml_html_modification_count, html_path, zip_path, main_text, text_file, text_length, n_char) %>%
  unique(.)

dbWriteTable(db_con, "works", works_df, overwrite = TRUE)
dbWriteTable(db_con, "authors", authors_df, overwrite = TRUE)
dbWriteTable(db_con, "texts", texts_df, overwrite = TRUE)
dbExecute(db_con, "EXPORT DATABASE 'db';")
dbDisconnect(db_con, shutdown = TRUE)

write_csv(meta_df, file = "./outputs/kaggle/aozora_corpus.csv")
write_csv(meta_df_en, file = "./outputs/kaggle/aozora_corpus_en.csv")

write_csv(meta_df[c("db_id", "main_text")], file = "./outputs/kaggle/main_text.csv")
write_csv(meta_df %>% select(., -"main_text"), file = "./outputs/kaggle/meta_info.csv")
write_csv(meta_df_en %>% select(., -"main_text"), file = "./outputs/kaggle/meta_info_en.csv")

save(meta_df, file = "./outputs/meta_df.RData")
save(meta_df_en, file = "./outputs/meta_df_en.RData")

#load("./outputs/meta_df.RData")
#load("./outputs/meta_df_en.RData")

File: azb.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

#Created on Mon Sep 18 07:22:07 PM EDT 2023 
#author: Ryan Hildebrandt, github.com/ryancahildebrandt

# imports
require "parallel"

require_relative "azb_utils"
require_relative "db_class"

db = AzbDB.new("outputs/kaggle/aozora_corpus.db")
db.load_metadata()
db.prepare_db()
db.load_static_arrays()

Parallel.map(db.meta_df, in_threads: 3) do |row|
#db.meta_df.each do |row|
  row["作品ID"] = row[0]
  row["XHTML/HTMLファイル符号化方式"] = db.encoding_key[row["XHTML/HTMLファイル符号化方式"]]
  row["テキストファイル符号化方式"] = db.encoding_key[row["テキストファイル符号化方式"]]
  row["本文"] = read_main_text(row["XHTML/HTMLファイルURL"], row["テキストファイルURL"], row["XHTML/HTMLファイル符号化方式"], row["テキストファイル符号化方式"])
  
  if row["本文"].to_s.empty?
    htm_local = File.exist?(row["XHTML/HTMLファイルURL"].gsub("https://www.aozora.gr.jp", "./aozorabunko")) ? row["XHTML/HTMLファイルURL"].gsub("https://www.aozora.gr.jp", "./aozorabunko") : ""
    txt_local = File.exist?(row["テキストファイルURL"].gsub("https://www.aozora.gr.jp", "./aozorabunko")) ? row["テキストファイルURL"].gsub("https://www.aozora.gr.jp", "./aozorabunko") : ""
    htm_remote = try_url(row["XHTML/HTMLファイルURL"]) ? row["XHTML/HTMLファイルURL"] : ""
    txt_remote = try_url(row["テキストファイルURL"]) ? row["テキストファイルURL"] : ""
    open("empties.txt", "a"){|f| f.puts "#{htm_local}, #{txt_local}, #{htm_remote}, #{txt_remote}"}
  end

  row["著者"] = "#{row["名"]}#{row["姓"]}"
  row["分類番号"] = row["分類番号"][/[0-9]{3}/]
  row["分類"] = db.genres_output.to_h[row["分類番号"]]
  
  row["振り仮名"]= extract_furigana(row["作品ID"], row["本文"])
  row["注釈"] = extract_notes(row["作品ID"], row["本文"])
  row["高低アクセント"] = extract_pitch_accents(row["作品ID"], row["本文"])
  row["本文"] = remove_notes(row["本文"])
  row["本文字数"] = row["本文"].length

  tokens = db.tagger.segment(row["本文"])
  row["常用"] = descriptives_from_key(tokens, db.joyo_key)
  row["日本語能力試験"] = descriptives_from_key(tokens, db.jlpt_key)
  row["出現頻度"] = descriptives_from_key(tokens, db.freq_key)
  row["通算難易度"] = aggregate_descriptives(row["常用"], row["日本語能力試験"], row["出現頻度"])

  db.authors_output.append([row["人物ID"], row["姓"], row["名"], row["姓読み"], row["名読み"], row["姓読みソート用"], row["名読みソート用"], row["姓ローマ字"], row["名ローマ字"], row["生年月日"], row["没年月日"], row["人物著作権フラグ"], row["著者"]])
  db.works_output.append([row["作品ID"], row["作品名"], row["作品名読み"], row["ソート用読み"], row["副題"], row["副題読み"], row["原題"], row["初出"], row["分類番号"], row["分類"], row["文字遣い種別"], row["作品著作権フラグ"], row["公開日"], row["最終更新日"], row["人物ID"], row["役割フラグ"], row["底本名1"], row["底本出版社名1"], row["底本初版発行年1"], row["入力に使用した版1"], row["校正に使用した版1"], row["底本の親本名1"], row["底本の親本出版社名1"], row["底本の親本初版発行年1"], row["底本名2"], row["底本出版社名2"], row["底本初版発行年2"], row["入力に使用した版2"], row["校正に使用した版2"], row["底本の親本名2"], row["底本の親本出版社名2"], row["底本の親本初版発行年2"], row["入力者"], row["校正者"]])
  db.texts_output.append([row["作品ID"], row["図書カードURL"], row["テキストファイルURL"], row["テキストファイル最終更新日"], row["テキストファイル符号化方式"], row["テキストファイル文字集合"], row["テキストファイル修正回数"], row["XHTML_HTMLファイルURL"], row["XHTML_HTMLファイル最終更新日"], row["XHTML_HTMLファイル符号化方式"], row["XHTML_HTMLファイル文字集合"], row["XHTML_HTMLファイル修正回数"], row["本文"], row["本文字数"]])
  db.difficulty_output.append([row["作品ID"], row["常用"], row["日本語能力試験"], row["出現頻度"], row["通算難易度"]].flatten)
  row["振り仮名"].each{|entry| db.furigana_output.append(entry)}
  row["注釈"].each{|entry| db.notes_output.append(entry)}
  row["高低アクセント"].each{|entry| db.accents_output.append(entry)}

  db.progress_bar.increment
end

db.export_db()
